command_line/generate_mask.py

```
@@ -32,108 +30,157 @@
+    """
+    Generate a pixel mask for each image in an experiment.
+
+    Use the masking parameters :param:`params` and an experiment in the experiment
+    list :param:`experiments` to define pixel masks for the associated imagesets.
+    The masks are generated using :mod:`dials.util.masking`.
+
+    The masks will be saved to disk at the location specified by
+    :attr:`params.output.mask`.  Optionally, if a path
+    :attr:`params.output.experiments` is set, a modified copy of :param:`experiments`
+    with the masks applied will be saved to that location.
+
+    Args:
+        experiments (:type:`dxtbx_model_ext.ExperimentList`): An experiment list
+            containing only one :type:`dxtbx_model_ext.Experiment` object.
+        params (:type:`libtbx.phil.scope_extract`): Masking parameters, having the
+            structure defined in :data:`dials.util.masking.phil_scope`.
+
+    Returns:
+        Tuple[:rtype:`scitbx_array_family_flex_ext.bool`]: A tuple containing the
+            generated pixel masks.
+        :rtype:`dxtbx_model_ext.ExperimentList` (optional): A copy of
+            :param:`experiments` with the masks applied.  Only returned if
+            :attr:`params.output.experiments` is set.
+    """
+    imagesets = experiments.imagesets()
+    if len(imagesets) > 1:
+        # Check beams (for resolution) and detectors are equivalent in each case
+        # otherwise the mask may not be appropriate across all imagesets
+        detectors = experiments.detectors()
+        beams = experiments.beams()
+        for d in detectors[1:]:
+            if not d.is_similar_to(detectors[0]):
+                raise ValueError(
+                    "Multiple imagesets are present, but their detector models differ."
+                )
+        for b in beams[1:]:
+            if not b.is_similar_to(beams[0]):
+                raise ValueError(
+                    "Multiple imagesets are present, but their beam models differ."
+                )
```

As above, I would argue that the user shouldn't be exposed to a full traceback here.

---

Agreed as above.

---

I know you'll hate me but I don't think sys.exit is the right thing either. It shows the message, but will also cause an unexpected exit if being used functionally.

Possible alternative: Raise `Sorry` which is an exception but also doesn't stacktrace, or throw something custom and catch it in the `run` function (don't want to hide any exceptions coming from underneath the `generate_mask` function).